### ISSUE TITLE
Error detecting end of heredocs

### ISSUE THREAD

**Issue by sriedel** [ruby/rdoc#727]
Trying to generate ri documentation for the childprocess gem version 0.9.0 results in an error:

    Parsing sources...
    Before reporting this, could you check that the file you're documenting
    has proper syntax:

      /home/sr/.rvm/rubies/ruby-2.6.3/bin/ruby -c lib/childprocess/tools/generator.rb

    RDoc is not a full Ruby parser and will fail when fed invalid ruby programs.

    The internal error was:

            (NoMethodError) undefined method `[]' for nil:NilClass

    /home/sr/.rvm/gems/ruby-2.6.3/gems/rdoc-6.1.1/lib/rdoc/parser/ripper_state_lex.rb:496:in `heredoc_end?'
            /home/sr/.rvm/gems/ruby-2.6.3/gems/rdoc-6.1.1/lib/rdoc/parser/ripper_state_lex.rb:473:in `get_heredoc_tk'
            /home/sr/.rvm/gems/ruby-2.6.3/gems/rdoc-6.1.1/lib/rdoc/parser/ripper_state_lex.rb:346:in `get_squashed_tk'
            /home/sr/.rvm/gems/ruby-2.6.3/gems/rdoc-6.1.1/lib/rdoc/parser/ripper_state_lex.rb:578:in `parse'
            /home/sr/.rvm/gems/ruby-2.6.3/gems/rdoc-6.1.1/lib/rdoc/parser/ruby.rb:180:in `initialize'
            /home/sr/.rvm/gems/ruby-2.6.3/gems/rdoc-6.1.1/lib/rdoc/parser.rb:191:in `new'
            /home/sr/.rvm/gems/ruby-2.6.3/gems/rdoc-6.1.1/lib/rdoc/parser.rb:191:in `for'
            /home/sr/.rvm/gems/ruby-2.6.3/gems/rdoc-6.1.1/lib/rdoc/rdoc.rb:360:in `parse_file'
            /home/sr/.rvm/gems/ruby-2.6.3/gems/rdoc-6.1.1/lib/rdoc/rdoc.rb:416:in `block in parse_files'
            /home/sr/.rvm/gems/ruby-2.6.3/gems/rdoc-6.1.1/lib/rdoc/rdoc.rb:414:in `map'
            /home/sr/.rvm/gems/ruby-2.6.3/gems/rdoc-6.1.1/lib/rdoc/rdoc.rb:414:in `parse_files'
            /home/sr/.rvm/gems/ruby-2.6.3/gems/rdoc-6.1.1/lib/rdoc/rdoc.rb:483:in `document'
            /home/sr/.rvm/gems/ruby-2.6.3/gems/rdoc-6.1.1/exe/rdoc:20:in `<top (required)>'
            /home/sr/.rvm/gems/ruby-2.6.3/bin/rdoc:23:in `load'
            /home/sr/.rvm/gems/ruby-2.6.3/bin/rdoc:23:in `<main>'
    undefined method `[]' for nil:NilClass
    
     $   /home/sr/.rvm/rubies/ruby-2.6.3/bin/ruby -c lib/childprocess/tools/generator.rb
     Syntax OK

It seems that the token passed to heredoc_end? is nil.

**Comment by joerixaop** [comment]
I've just encountered this issue as well, after some debugging I discovered that if the file uses a DOS line ending (`\r\n`) then the heredoc ending will contain the `\r`, while being compared to the heredoc name without `\r` (at least on unixy systems, and using Ruby 2.6.5). Obviously those don't compare equal so it keeps looking for tokens that could be the end of the heredoc until it runs out.

Running out of tokens should probably report in an error giving some more detail of what went wrong "Couldn't find the end of heredoc X, that started on line n", in addition the type of line endings shouldn't really matter.

Seems like this was fixed by commit 5f9603f35d8e520c761015810c005e4a5beb97c3
